<style>
	h1 {
	--	background-color: grey;
		color: yellow
		border: 3px solid grey

	}
	p {
		background-color: grey;
		color: white
		border: 3px solid grey

	}
</style>
<h1> Chapter 5 Functions: Part 1</h1>

<p style="margin-left:40px">1. 5.1 Introduction to Functions</p>

<p style="margin-left:40px">2. 5.2 Defining and Calling a Void Function</p>

<p style="margin-left:40px">3. 5.3 Designing a Program to Use Functions</p>

<p style="margin-left:40px">4. 5.4 <mark> Local Variables </mark></p>

<p style="margin-left:40px">5. 5.5 <mark> Passing Arguments to Functions </mark></p>

<p style="margin-left:40px">6. 5.6 Global Variables and Global Constants</p>

<p style="margin-left:40px">7. 5.7 Introduction to <mark>Value-Returning Functions </mark>: Generating Random Numbers</p>

<p style="margin-left:40px">8. 5.8 Writing Your Own Value-Returning Functions</p>

<p style="margin-left:40px">9. 5.9 The math Module</p>

<p style="margin-left:40px">10. 5.10 <mark> Storing Functions in Modules <mark> </p>

<p style="margin-left:40px">11. 5.11 Turtle Graphics: Modularizing Code with Functions</p>


***

## <p style="background-color: grey; color: white "> ◼️  functions </p>
# <p style="font-family: Arial; font-size:1.4em;color:red;"> 1. Basic Examples of Functions</p>


<style>
	h3{
		background-color: grey
		color: white
	}
</style>

<h3> First example of function </h3>

- Run all the following code cells and check out the result of code segments

#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #1.0 </p>

#### Run all the cells and figure out the basic usage of while loop structure

In [ ]:
def fn():
	return

if __name__ == '__main__':
	fn( )



In [ ]:
# Function to return 0 or 1
# if the number is even, return 0, otherwise 1
#  checknum ( )

def checknum(n):
	if n%2 == 0:
		return 0
	else:
		return 1

userinput = int(input())
retval = checknum(userinput)
print(retval)



3
1


In [ ]:
def printlst(l):
	average = sum(l) / len(l)
	for v in l:
		if ( average > v):
			print(l)

lst = [1,2,3,4,5,6,7]
printlst(lst) # print all numbers less than avg
print (lst)


[1, 2, 3, 4, 5, 6, 7]
[1, 2, 3, 4, 5, 6, 7]
[1, 2, 3, 4, 5, 6, 7]
[1, 2, 3, 4, 5, 6, 7]


In [ ]:

def 	getinput():
	val = int(input())
	return val

userval = getinput()
print (userval)




12
12


### How are the functions called?  
#### Compare the below two code segments

In [ ]:
def 	getinput():
	val = int(input())
	return val

def 	main():
	userval = getinput()
	print (userval)


if __name__ == '__main__':
	main()



12
12


In [ ]:
def 	getinput():
	val = int(input())
	return val

def 	main():
	userval = getinput()
	print (userval)


#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #1.1 </p>

####  Edit the above code cell to run main() function

```
def 	getinput():
	val = int(input())
	return val

def 	main():
	userval = getinput()
	print (userval)
```

In [ ]:
# complete the code to run the main function

def     getinput():
    val = int(input())
    return val

def     main():
    userval = getinput()
    print (userval)

if __name__ == '__main__':
    main()

13
13


#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #1.2 </p>

####  Check out this code. This code is working without main() function.
#### Explain how this code is working properly without main() function

```
def 	getinput():
	val = int(input())
	return val

userval = getinput()
print (userval)
```

The function main() exectues becasue it is being called by when it is assigned to the variable 'userval'.

In [ ]:
def 	getinput():
	val = int(input())
	return val

userval = getinput()
print (userval)


33
33


## Don't forget to <font color='red'> **"commit"** </font> your work or <font color='red'> **"save"** </font> in Google Colab
## before moving to the next section.

***

## <p style="background-color: grey; color: white "> ◼️  Local Variables </p>
# <p style="font-family: Arial; font-size:1.4em;color:red;"> 2. Local Variables </p>


### Check out the variable "userval" after the function call "getinput()"

In [ ]:
def 	getinput():
	val = int(input())
	return val

def 	main():
	userval = getinput()
	print (userval)
	# print (val)

if __name__ == '__main__':
	main()



11
11


In [ ]:
def getsum(n1, n2):
	val = n1 + n2
	return val

def main():
	a = 10; b = 20
	userval = getsum(a, b)
	print (userval)

if __name__ == '__main__':
	main()


30


<style>
	p {
		background-color: grey;
		color: white;
		border: 3px solid grey

	}
</style>

<p> Making Changes to Parameters </p>


#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #2.1 </p>

####  Check out this code. After the function call "getinput()" at the line 8 in the main function,  
#### show the  values "a" and "b"


In [ ]:

def getinput(n1, n2):
	n1 = input()
	n2 = input()
	print ("The value of parameter n1 and n2 are {0} and {1}".format(n1, n2))

def main():
	a = 0; b = 0
	getinput(a, b)
	print ("The value of a and b are {0} and {1}".format(a, b))

if __name__ == '__main__':
	main()


4
5
The value of parameter n1 and n2 are 4 and 5
The value of a and b are 0 and 0


### <mark> Explain why both values are 0 </mark>

- The values 'a' and 'b' are immutable integers. Even though the values are hardcoded and predefined, they remain unchanged when calling the function 'getinput()' because 'a' and 'b' are immutable.

---
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #2.2 </p>

####  Check out this code. After the function call "getinput()" in the main function,  
#### show the  values "list1". Explain the result.
### In problem 2.1, there were no changes even though the values were altered in the function getinput().
### In this code, when you assign the values to the list "l1", these changes affect to the main function.


<style>
	p {
		background-color: grey;
		color: white;
		border: 3px solid grey

	}
</style>

<p> Making Changes to Parameters 2 </p>


In [ ]:

def getinput(l1):
	l1.append(input())
	l1.append(input())
	l1.append(input())
	print ("The value of parameter l1 ", l1)

def main():
	list1 = []
	getinput(list1)
	print ("The value of list1", list1);

if __name__ == '__main__':
	main()


a b c
d
e
The value of parameter l1  ['a b c', 'd', 'e']
The value of list1 ['a b c', 'd', 'e']


## Answer the question 2.2 here
- Lists are mutable whereas integers are immutable.

---
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #2.3 </p>

####  Check out this code.
####  Can we return multiple values in the function?


<style>
	p {
		background-color: grey;
		color: white;
		border: 3px solid grey

	}
</style>

<p> Returning multiple values</p>


In [ ]:

def getinput():
	val1 = input()
	val2 = input()
	return val1, val2

def main():
	num1, num2 = getinput()
	print ("The value of num1 and num2 are {0} and {1}".format(num1, num2))

if __name__ == '__main__':
	main()


33
12
The value of num1 and num2 are 33 and 12


## Answer the question 2.3 here
- Yes

---
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #2.4 </p>

####  Print the local variable that was used in the function getsum() in the main() function.
####  Can we print the same value ? Errors? Why?
#### Suggest how we can print the 'total' of the function getsum() in the main()


In [2]:
def getsum(n1, n2):
    total = n1 + n2
    print(f'getsum n1 and n2: {n1}, {n2}')
    print(f'Total is {total}')

def main():
    total = 0
    num1 = 10
    num2 = 20
    getsum(num1, num2)
    print(f'main num1 and num2 after calling getsum: {num1}, {num2}')
    print(f'Total is {total}')

if __name__ == '__main__':
    main()

getsum n1 and n2: 10, 20
Total is 30
main num1 and num2 after calling getsum: 10, 20
Total is 0


## Answer the question 2.4 here
- num1 and num2 are immutable; therefore, the the getsum function is irrelevant because it cannot change/mutate immutable variables.
- to print 'total' of the function getsum() within the main() function you would need to change num1 and num2 in main() to different variables e.g. n1 and n2.

---
## Don't forget to <font color='red'> **"commit"** </font> your work or <font color='red'> **"save"** </font> in Google Colab
## before moving to the next section.

<style>
	p {
		background-color: grey;
		color: white;
		border: 3px solid grey

	}
</style>

<p> Keyword Arguments </p>


***
## <p style="background-color: grey; color: white "> ◼️  Keyworded Arguments </p>
# <p style="font-family: Arial; font-size:1.4em;color:red;"> 3. Keyworded Arguments</p>


#### The parameter order can be changed if the keyworded arguments are used

In [3]:

def getsum(n1, n2):
	print (f'The first parameter n1 is {n1}')
	print (f'The second parameter n2 is {n2}')
	total = n1 + n2
	print (f'The total is {total}')
	return total

def main():
	total = getsum(n2=100, n1=200)	# giving the second parameter first
	print (f'The value of total is {total}')

if __name__ == '__main__':
	main()

The first parameter n1 is 200
The second parameter n2 is 100
The total is 300
The value of total is 300


#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #3.1 </p>

### The function setListValues() will work to make the mylst that has the values 100, 200 and 300
####  Complete the function call to make the list as following
```
mylst = [100, 200, 300]


In [18]:
def setListValues(mylst, first_element, second_element, third_element):
    mylst.append(first_element)
    mylst.append(second_element)
    mylst.append(third_element)

def main():
    mylst=[]
    setListValues(mylst,
                  third_element= 300,
                  second_element= 200,
                  first_element= 100)
    print(mylst)

if __name__ == '__main__':
    main()


[100, 200, 300]


### Follow up Question

### Is the list "mylst" in function "setListValues()" as the one in "main"?
----

### Your Answer :
- Yes, 'mylst' is the same if it is passed from main() to setListValues().

## Don't forget to <font color='red'> **"commit"** </font> your work or <font color='red'> **"save"** </font> in Google Colab
## before moving to the next section.

***
## <p style="background-color: grey; color: white "> ◼️  Global Variables </p>
# <p style="font-family: Arial; font-size:1.4em;color:red;"> 4. Global Variables</p>


<style>
	p {
		background-color: grey;
		color: white;
		border: 3px solid grey

	}
</style>

<p> Global Variables </p>

Run the below code and check out the values <mark> gvar </mark>


In [19]:
gvar = 100

def myfunction1():
	print ('Inside the function ', gvar)


def main():
	global gvar
	print ('Before call myfunction1 ', gvar)
	gvar += 11

	myfunction1()
	print ('After call myfunction1 ', gvar)

if __name__ == '__main__':
	main()

Before call myfunction1  100
Inside the function  111
After call myfunction1  111


#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #4.1 </p>

####  Explain the printed the value of "gvar"  


In [20]:
gvar = 100

def myfunction1():
    global gvar
    gvar += 1

def main():
	global gvar
	print ('Before call myfunction1 ', gvar)
	myfunction1()
	print ('After call myfunction1 ', gvar)
	gvar += 1
	print ('After increasing by 1 in main function', gvar)

if __name__ == '__main__':
	main()

Before call myfunction1  100
After call myfunction1  101
After increasing by 1 in main function 102


- gvar begins as 100. When called by main()  it becomes 101 because it is called by myfunctions1() that adds 1 to 100 to yield 101. After myfuntion1() is run there is another +1 to yeield 102, which is the final value of gvar.

#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #4.2 </p>

####  Explain the printed the value of "gvar" is 100 even though it has been increased by 1 in the myfuntion1()


In [21]:
gvar = 100

def myfunction1():
    gvar = 0
    gvar += 1

def main():
	global gvar
	print ('Before call myfunction1 ', gvar)
	myfunction1()
	print ('After call myfunction1 ', gvar)

if __name__ == '__main__':
	main()

Before call myfunction1  100
After call myfunction1  100


- gvar remains 100 because the 'gvar' in myfunction1() is not a global variable but a local variable. myfunction1() is missing the global declaration to make 'gvar' a global variable rather than a local variable.

#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #4.3 </p>

####  Explain why this code encounters error
####  Fix the Errors and show the complete code


In [22]:
gvar = 100

def myfunction1():
    gvar += 1

def main():
	global gvar
	print ('Before call myfunction1 ', gvar)
	myfunction1()
	print ('After call myfunction1 ', gvar)

if __name__ == '__main__':
	main()

Before call myfunction1  100


UnboundLocalError: local variable 'gvar' referenced before assignment

- 'gvar' is not defined as a variable in myfunction1(). It can be assigned either as a local or global variable.

In [23]:
gvar = 100

def myfunction1():
    global gvar  # must define 'gvar' as a global variable
    gvar += 1

def main():
	global gvar
	print ('Before call myfunction1 ', gvar)
	myfunction1()
	print ('After call myfunction1 ', gvar)

if __name__ == '__main__':
	main()

Before call myfunction1  100
After call myfunction1  101


## Please <font color='red'> **"commit"** </font> your work or <font color='red'> **"save"** </font> in Google Colab
## before moving to the next section.

***
## <p style="background-color: grey; color: white "> ◼️  Arbitrary Arguments and Keyword Arguments</p>
# <p style="font-family: Arial; font-size:1.4em;color:red;"> 5. Arbitrary Arguments and Keyword Arguments</p>


## Arbitrary Argument

### An arbitrary argument in Python allows a function to accept an unspecified number of arguments using *args for non-keyword arguments or **kwargs for keyword arguments

### Arbitrary Arguments (*args):
```
* is used to allow a function to accept an arbitrary number of positional arguments.
```

Basic Example

In [24]:
def greet(*args):
    for name in args:
        print(f"Hello, {name}!")

greet("Alice", "Bob", "Charlie")


Hello, Alice!
Hello, Bob!
Hello, Charlie!


#### Understanding the function type()

In [25]:
number = 10
type(number)

int

### Check the data type for the arbitrary argument


#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #5.1 </p>

####  Check the printed data type.  
####  Type the printed data type in the below


In [27]:
def myfunction(*values):
    print(type(values))
    print (values)
    for v in values:
        print (v, end='\t')
        print(type(v))

n1 = 10; n2 = 20; n3=30;
myfunction(n1, n2, n3)

<class 'tuple'>
(10, 20, 30)
10	<class 'int'>
20	<class 'int'>
30	<class 'int'>


#### <mark> Type your answer here </mark>
- What is the data type of values in the function?
    - tuple, because the '*' accepts any number of arguments and then packs them into a tuple

***
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #5.2 </p>

####  Check the printed data type.  
####  Type the printed data type in the below


In [39]:
def myfunction(*args):
    print(type(args), 'It has been printed in the funtion myfuntion()') # type the printed data type in the below cell
    print ('The value of args is ', args)
    for v in args:
        print (v, end='\t')
    print('\n',type(v))
    print(' ')
def main():
    values = [10, 20, 30]
    print (type(values), 'It has been printed in the main()')    # type the printed data type in the below cell
    myfunction(values)

if __name__ == '__main__':
    main()


<class 'list'> It has been printed in the main()
<class 'tuple'> It has been printed in the funtion myfuntion()
The value of args is  ([10, 20, 30],)
[10, 20, 30]	
 <class 'list'>
 


#### <mark> Type your answer here </mark>

### Are the two printed data types same?: No, the list called 'values' from main() becomes a tuple in myfunction()

- What is the data type printed in main()?
    - list
- What is the data type printed in myfunction()?
    - tuple
- How many elements are there in the tuple "args"?
    - one
- What is the data type of <mark> element </mark> in the tuple "args"?
    - list
- How many elements are there in the list "values"?
    - 3
-How many iterations are occured in the for loop of myfuntion()? Why? Explain it.
    - 1, because there is one list within the tuple

***
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #5.3 </p>

####  Check the printed data type and clarify the difference printed output values between two function calls
####  


In [4]:
def myfunction(*values):
    print(type(values)) # type the printed data type in the below cell
    print (values)
    for v in values:
        print (v, end='\t')
    print()

def main():
    values = [10, 20, 30]
    print(f'type for values in main(): {type(values)}')
    myfunction(values)          # It prints (10, 20, 30)    Why ?

    val1 = 10 ; val2 = 20; val3 = 30;
    print(f'type for val1 in main(): {type(val1)}')
    myfunction(val1, val2, val3)        # It prints 10 20 30 ; Why?
if __name__ == '__main__':
    main()


type for values in main(): <class 'list'>
<class 'tuple'>
([10, 20, 30],)
[10, 20, 30]	
type for val1 in main(): <class 'int'>
<class 'tuple'>
(10, 20, 30)
10	20	30	


#### <mark> Type your answer here </mark>

### Are the two printed data types same?

- What is the data type printed in main()?
    - list for the variable 'values'
    - int for the varaibles e.g. 'val1'
- What is the data type printed in myfunction()?
    - tuple for both the list and the integers because the '*' packs the arguments into a tuple as is the nature of '*' aka *args
- How many elements are there in the tuple 'values' at each call myfunction() ?
    - 1, regardless if the values being pulled (aka arguments) is the list or integers
- How many iterations are occured in the for-loop? why?
    - 1, for the list within the tuple
    - 3, for the three integers within the tuple

***
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #5.4 </p>

####  Clarify the differences between two functions "myfn1" and "myfn2"
####  Answer in the below markdown cell.


In [6]:
def myfn1(*values):
    print (type(values))
    for v in values:
        print (v, end='\t')

def myfn2(values):
    print (type(values))
    for v in values:
        print (v, end='\t')

values = [10, 20, 30]
myfn1(values)
print()
myfn2(values)

<class 'tuple'>
[10, 20, 30]	
<class 'list'>
10	20	30	

#### <mark> Type your answer here </mark>

### Are the two printed data types same?

- Are the printed values are same? If not, why?
    - No, because * (*args) is a parameter that takes arguments and packs them into a tuple

## Unpacking Operator:

### <mark> * </mark> is used to unpack elements from a list or tuple into separate variables or to pass multiple arguments to a function.

In [8]:
# check out the difference between two print statements
numbers = [10, 20, 30]
print (numbers)
print (*numbers)
print (numbers)


[10, 20, 30]
10 20 30
[10, 20, 30]


***
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #5.5 </p>

####  Clarify the differences of the printed output between two functions "myfn1" and "myfn2"
####


In [16]:
def myfn1(n1, n2, n3):
    print (n1, n2, n3)
    print(type(n1))

def myfn2(val):
    print (val)
    for v in val:
        print (v, end=' ')
        print(type(v))
    print()

numbers = [10, 20, 30]
myfn1(*numbers)
myfn2(numbers)

10 20 30
<class 'int'>
[10, 20, 30]
10 <class 'int'>
20 <class 'int'>
30 <class 'int'>



#### <mark> Type your answer here </mark>

- Are the two printed output same?
    - Yes, while one is a list and the other an unpacked tuple, the output is the same.

- Are the printed values are same? If not, why?
    - Yes, both are class 'int' aka integer

- Is it possible to call function myfn1(numbers)? Clarify your answer.
    - No, because myfn1 is expecting 3 arguments.
- Is it possible to call function myfn2(*numbers)? Clarify your answer.
    - No, myfn2 is expecting one arguments and therefore '*' will not work as it will unpack 3 arguments.

***

# Keyword Arguments (**args):
```
* is used to allow a function to accept an arbitrary number of positional arguments.
```

Basic Example

***
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #5.5 </p>

####  Run the below code cell  
####  Answer in the below markdown cell.


### **kwargs
## Run the following code and learn about the keyword argument

In [17]:
def myFun(**kwargs):
	print (type(kwargs), 'The type of kwargs')
	print (isinstance(kwargs, dict), '<--- If this is true, kwargs is a dictionary ')
	for key, value in kwargs.items():
		print ("%s -> %s" %(key, value))


myFun(first ='California', mid ='Community', last='College')


<class 'dict'> The type of kwargs
True <--- If this is true, kwargs is a dictionary 
first -> California
mid -> Community
last -> College



-  What is the type of "kwargs" ? <br>
-  Is it true "isinstance(kwargs, dict) ?  <br>

#### <mark> Type your answer here </mark>
- What is the type of "kwargs"?
    - kwargs stands for keyword arguments and is represented as '**' to create a dictionary. The 'key' in keyword implies key-value pairs for a dictionary.
- Is it true "isinstance(kwargs, dict)?
    - Yes

#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #5.6 </p>

####  Call the function myFun( ) with the following code
``` myFun(mid ='Community', last='College', first ='California') ```

#### And explain the output. Was output printed from "first" key value or "mid" key value?


In [20]:
def myFun(**kwargs):
	for key, value in kwargs.items():
		print ("%s -> %s" %(key, value))

myFun(mid ='Community', last='College', first ='California')


mid -> Community
last -> College
first -> California


- No, dictionaries preserve the order of argumenents being passed

## Call function with the single value and keyword argument
```
myFun( normalvalue, **keyword_argument)
```

In [21]:
def myFun(first, **kwargs):
	print (first)
	for key, value in kwargs.items():
		print ("%s -> %s" %(key, value))

# call myfun() with separate key=value pairs
myFun('Apple', mid ='Pear', last='Orange')

# Or
kwargs = {'mid':'Pear', 'last':'Orange'}
myFun('Apple',**kwargs)



Apple
mid -> Pear
last -> Orange
Apple
mid -> Pear
last -> Orange


#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #5.7 </p>

####  Call the function myFun( ) with the following code
```
	# Call 1
	myFun("Java", "C++", "Python")
	myFun(arg2="C++", arg3="Python", arg1="Java")

	# Call 2
	args = ("Java", "C++", "Python")
	myFun(*args)

	# Call 3
	kwargs = {"arg1" : "Java", "arg2" : "C++", "arg3" : "Python"}
	myFun(**kwargs)
```
#### And explain the output. All outputs are same?


## Call function(a, b, c) with
- normal value argument: -----   function(arg1, arg2, arg3)
- arbitrary argument:-----       function(*args)
- keyword argument:-----         function(**kargs)
```
def myFun( a, b, c):
```

In [22]:
def myFun(arg1, arg2, arg3):
	print("arg1:", arg1, end='\t')
	print("arg2:", arg2, end='\t')
	print("arg3:", arg3, end='\n')

def main():
	# Call 1
	myFun("Java", "C++", "Python")
	myFun(arg2="C++", arg3="Python", arg1="Java")

	# Call 2
	args = ("Java", "C++", "Python")
	myFun(*args)
	# Call 3
	kwargs = {"arg1" : "Java", "arg2" : "C++", "arg3" : "Python"}
	myFun(**kwargs)


if __name__ == '__main__':
    main()


arg1: Java	arg2: C++	arg3: Python
arg1: Java	arg2: C++	arg3: Python
arg1: Java	arg2: C++	arg3: Python
arg1: Java	arg2: C++	arg3: Python


#### <mark> Type your answer here </mark>

### All the output are same? Why? Explain it.
- Yes
    - Call 1 uses assigned variables to maintain the order by assignment even though the arguments are not in the proper order.
    - Call 2 the arbritrary arguments are in the proper order
    - call 3 the key-value piars in **kwargs is in the correct order


## Other usefule expressions with arbitarary argument

#### Run the followig code cells and learn the arbitrary argument

<style>
	p {
		--background-color: grey;
		color: white;
		border: 3px solid grey

	}
</style>

<p> Using the arbitrary arguments to save the returned value</p>



In [28]:
def retEven(lst):
	evenlst = []
	for v in lst:
		if v % 2 == 0:
			evenlst.append(v)
	return evenlst

lst = [1,2,3,4,5,6,8]
first, *others, last = retEven(lst)
print (first)
print (last)
print (others)

# * packs all the items except for two items assigned to first and last.
# Also, there must be 3 or more elements in 'lst'


2
8
[4, 6]


<style>
	h1 {
		background-color: grey;
		color: Yellow;
		border: 3px solid grey;
		-- font-size: xx-large;
	}
	p {
		background-color: grey;
		color: white;
		border: 3px solid grey;
		font-size: xx-large
	}
</style>
<p> Example: *args </p>


In [29]:

def printstring(*string):
	print (string)
	for v in string:
		print ('The value v is ', v, '\tThe type of v is ', type(v) )

	print ()

stringvalue = "Python Programming"
printstring(stringvalue)

integerlist = [1,2,3,4,5]
printstring(integerlist)

printstring(1,2,3,4,5)

('Python Programming',)
The value v is  Python Programming 	The type of v is  <class 'str'>

([1, 2, 3, 4, 5],)
The value v is  [1, 2, 3, 4, 5] 	The type of v is  <class 'list'>

(1, 2, 3, 4, 5)
The value v is  1 	The type of v is  <class 'int'>
The value v is  2 	The type of v is  <class 'int'>
The value v is  3 	The type of v is  <class 'int'>
The value v is  4 	The type of v is  <class 'int'>
The value v is  5 	The type of v is  <class 'int'>



## Please <font color='red'> **"commit"** </font> your work or <font color='red'> **"save"** </font> in Google Colab
## before moving to the next section.

***
# <p style="background-color: grey; color: white "> ◼️  Section 6 </p>
# <p style="font-family: Arial; font-size:1.4em;color:red;"> 6. Yield </p>


<style>
	h1 {
		background-color: grey;
		color: Yellow;
		border: 3px solid grey;
		-- font-size: xx-large;
	}
	p {
		background-color: grey;
		color: white;
		border: 3px solid grey;
	}
</style>
<h1 font-size:40px> 6.1 iterator </h1>



## Understanding Iterator

#### Run the followig code cells and learn the Iterator concept

<style>
	h3 {
		background-color: black;
		color: white;
		border: 1px solid grey;
		font-size: 30px;
	}
	p {
		background-color: #336666;
		color: white;
		border: 1px solid grey;
		font-size: 24;
	}
</style>

<h3> Iterator </h3>
<p>

An iterator in Python is an object that enables you to traverse through a collection of elements, such as a list, tuple, or dictionary, one element at a time. It implements two main methods:
```
__iter__(): Returns the iterator object itself. This is required for an object to be an iterator.
__next__(): Returns the next item from the collection. When there are no more items, it raises a StopIteration exception.
```
</p>

#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #6.1 </p>

<style>
	p {
		background-color: black;
		color: white;
		border: 1px solid grey;
		font-size: 30px;
	}
</style>
<p> Run the following code cells and answer the questions </p>  


#### What is the type of "myiter"?
- Your answer: 👉

In [40]:
mylst = [10, 20, 30]
myiter = iter(mylst)
print(myiter)
print(iter(mylst))
type(myiter)

list_iterator

#### What is the result of the following code?
``` myiter.__next__() ```
- Your answer: Here 👉
    - the next item in the collection. Since it is being called for the first time, it would be the first item, 10

In [33]:
myiter.__next__()

10

#### What is the result of the following code?
```
    myiter.__next__()  
    myiter.__next__()
```
- Your answer: Here 👉

In [34]:
print (myiter.__next__())
print (myiter.__next__())

20
30


#### If you run this code one more, you will see error. Show the error message. Why does it cause an error?
```
    myiter.__next__()  
```
- Your answer: Here  👉

In [35]:
print (myiter.__next__())

StopIteration: 

#### next() function is the same as ``` __next__() ```
```
    iter_obj = iter(mylst)
    next(iter_obj)
```

In [36]:
mylst = [1,2,3]
itobj= iter(mylst)
print (next(itobj))
print (next(itobj))
print (next(itobj))

1
2
3


***
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #6.2: Loop Structures with Iterator </p>

<style>
	p {
		background-color: black;
		color: white;
		border: 1px solid grey;
		font-size: 30px;
	}
</style>
<p> Run the following code cells and answer the questions </p>  


#### Compare two loop structures with iterator, for loop and while loop.

<style>
	h5 {
		background-color: black;
		color: white;
		border: 1px solid grey;
		font-size: 30px;
	}
	p {
		background-color: #336666;
		color: white;
		border: 1px solid grey;
		font-size: 24;
	}
</style>
##### for loop
<p>
Using Iterators in Loops <br>
Most commonly, you use iterators in a for loop, which automatically handles the StopIteration exception:
</p>

In [37]:
mylst = [10, 20 , 30]
myiter = iter(mylst)
for v in myiter:
    print (v, end='\t')

10	20	30	

<style>
	h5 {
		background-color: black;
		color: white;
		border: 1px solid grey;
		font-size: 30px;
	}
	p {
		background-color: #336666;
		color: white;
		border: 1px solid grey;
		font-size: 24;
	}
</style>
##### while loop

In [38]:
mylst = [10, 20, 30]
iter_obj = iter(mylst)
while True:
    try:
        print (next(iter_obj), end= '\t')
    except StopIteration:
        break


10	20	30	

### Compare two loop structures. What is the difference between two loops to use the iterator for traversing all the elements in the list?
#### Your Answer: Here 👉

- The for loop interates 3 times as there are 3 elements in myiter
- The while loop interates until the exception is raised, 'StopIteration', which occurs when there are no more elements in myiter

<style>
	h1 {
		background-color: grey;
		color: Yellow;
		border: 3px solid grey;
		font-size: xx-large;
	}
	p {
		background-color: grey;
		color: white;
		font-size: 24px;
		border: 3px solid grey;
	}
</style>
# <h1 font-size:40px> 6.2 Generator </h1>


## Concepts of Generator
<p>
A generator in Python is a special type of iterator that allows you to iterate over a sequence of values lazily, meaning it generates values on the fly and only when needed. <br> This makes generators memory-efficient, especially when dealing with large datasets, because they do not require all the values to be stored in memory at once.
</p>

***
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #6.3 </p>

<style>
	p {
		background-color: black;
		color: white;
		border: 1px solid grey;
		font-size: 30px;
	}
</style>
<p> Run the following code cells and answer the questions </p>  


#### Run this code and check out the output. Show the type of "gen"
```
def mygen():
    yield 1

gen = mygen()
print (type(gen))
```
##### Your answer: Here 👉

- gen is of the class 'generator'

***

#### Run this code and check out the output. What are the value from "next(gen)"?
```
def mygen():
    yield 10
    yield 20
    yield 30

gen = mygen()
print (next(gen))
print (next(gen))
print (next(gen))
```
##### Your answer: Here 👉

- 10, 20, 30

***

In [41]:
def mygen():
    yield 1

gen = mygen()
print (type(gen))

<class 'generator'>


In [42]:
def mygen():
    yield 10
    yield 20
    yield 30

gen = mygen()
print (next(gen))
print (next(gen))
print (next(gen))

10
20
30


#### Run this code and check out the output. What are the value from "next(gen)"?
```
def mygen():
    yield 10
    yield 20
    yield 30

gen = mygen()
for v in gen:
    print (v, end='\t')
```
##### Your answer: Here 👉

- 10 20 30

***

In [43]:
def mygen():
    yield 10
    yield 20
    yield 30

gen = mygen()
for v in gen:
    print (v, end='\t')

10	20	30	

#### Run this code and check out the output. What is the value of the last next(gen) ?
```
def mygen():
    yield 10
    yield 20
    yield 30

gen = mygen()
while True:
    try:
        print (next(gen))
    except StopIteration:
        break
```
##### Your answer: Here 👉

- 30 is the last value; although, technically the last value is a 'StopIteration' error.

***

In [44]:
def mygen():
    yield 10
    yield 20
    yield 30

gen = mygen()
while True:
    try:
        print (next(gen))
    except StopIteration:
        break

10
20
30


<style>
	h1 {
		background-color: grey;
		color: Yellow;
		border: 3px solid grey;
		font-size: xx-large;
	}
	p {
		background-color: grey;
		color: white;
		font-size: 24px;
		border: 3px solid grey;
	}
</style>
# <h1 font-size:40px> 6.3 yield </h1>


## yield statement
<p>
yield is a keyword used in a function to turn it into a generator. When a function contains a yield statement, it does not execute like a normal function. Instead, it returns a generator object that can be iterated over.
</p>

***
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #6.4 </p>

<style>
	p {
		background-color: black;
		color: white;
		border: 1px solid grey;
		font-size: 30px;
	}
</style>
<p> Run the following code cells and answer the questions </p>  


<style>
	p {
		--background-color: grey;
		color: white;
		border: 3px solid grey

	}
</style>

<p> Example of yield statement </p>
<p> Make a function to return the even number in the list </p>



In [46]:
def findEvenNumber(lst):
	for v in lst:
		if ( v % 2 == 0):
			yield v

def main():
	mygen = findEvenNumber(range(10))
	val = next(mygen)
	print (val, '--')
	for i in mygen:
		print (i)

if __name__ == '__main__':
	main()

0 --
2
4
6
8


<style>
	p {
		--background-color: grey;
		color: white;
		border: 3px solid grey

	}
</style>

<p> Make a function to return the generator for the Fibonacci number </p>



In [49]:

def fibo(N):
	n1, n2 = 0, 1
	i = 0
	while ( i <= N):
		yield n1
		n3 = n1 + n2
		n1 = n2
		n2 = n3
		i += 1

gen = fibo(1)
print(f'gen: {gen}')
for v in gen:
	print (v)


gen: <generator object fibo at 0x784b4780a500>
0
1


<style>
	h1 {
		background-color: grey;
		color: Yellow;
		border: 3px solid grey;
		-- font-size: xx-large;
	}
	p {
		background-color: grey;
		color: white;
		border: 3px solid grey;
		font-size: xx-large
	}
</style>
<p> Example: yield </p>


In [50]:

def consonant(string):
	vowels = ['a', 'e', 'i','o', 'u']
	for v in string:
		if v not in vowels:
			yield v


strval = 'Python Programming'
for v in consonant(strval):
	print (v, end=', ' )

P, y, t, h, n,  , P, r, g, r, m, m, n, g, 

<style>
	p {
		background-color: black;
		color: white;
		border: 1px solid grey;
		font-size: 30px;
	}
</style>
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #6.5 </p>

<p> Make a function to return the generator containing the first half of the list


In [57]:
def halfoflist(mylst):
    # mid_way = len(mylst) // 2
    # yield mylst[0:mid_way]
    yield mylst[0:len(mylst) // 2]


def main():
    numbers = [1,2,3,4,5,6,7,8,9,10]
    mygen = halfoflist(numbers)

    for v in mygen:
        print (v, end ='\t')

if __name__ == '__main__':
    main()


[1, 2, 3, 4, 5]	

***
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #6.6 </p>

<style>
	p {
		background-color: black;
		color: white;
		border: 1px solid grey;
		font-size: 30px;
	}
</style>
<p> Make a function that returns the generator containing elements greater than their immediate left. </p>
- Example numbers = [1, 3, 2, 4, 5, 0] <br>
- Generator for [3,4,5]


In [68]:
### Code your program here
def mylist(num):
    for n in range(1, len(num)):
        if num[n] > num[n-1]:
            yield num[n]

numbers = [1, 3, 2, 4, 5, 0]
gen = mylist(numbers)
# number_list = []
# for i in gen:
#     number_list.append(i)
# print(number_list)

new_numbers = list(gen)
print(new_numbers)
print(type(new_numbers))

[3, 4, 5]
<class 'list'>


<style>
	h1 {
		background-color: grey;
		color: Yellow;
		border: 3px solid grey;
		font-size: 60px;
	}
	p {
		background-color: grey;
		color: white;
		font-size: 20px;
		border: 3px solid grey;
	}
</style>
# <h1 font-size:60px> 6.4 lambda </h1>


## Lambda Function
<p>
A lambda function in Python is a small, anonymous function defined with the lambda keyword. Unlike a regular function that is defined using the def keyword, a lambda function can have any number of arguments but only one expression. The result of that expression is implicitly returned.
</p>

Syntax <br>
<code>
lambda arguments: expression <br>

Code Example: <br>
<code>
add_ten = lambda x: x + 10 <br>
print(add_ten(5))  # Output: 15


***
<style>
	p {
		background-color: black;
		color: white;
		border: 1px solid grey;
		font-size: 30px;
	}
</style>
<p> Run the following code cells and check out the results </p>  


#### Lambda Function Example: Add 10

In [69]:

lfn = lambda x : x + 10

a = lfn(20)
print (a)

30


### Lambda that returns x squared

In [70]:
squared = lambda x : x * x

print (squared(10))

100


### Lambda that returns a boolean as a result of x > y

In [71]:
greater = lambda x, y : x > y

print (greater(10, 20))
print (greater(20, 10))


False
True


### Function that returns a lambda

In [75]:
def retlam( value):
	return lambda parameter : parameter * value

mylambda = retlam(10)
print (mylambda(20))

mylambda = retlam(20)
print (mylambda(20))

200
400


### Lambda Examples

In [76]:
import random

lst = [ random.randint(0,100) for i in range(5)]
print (lst)

fmin = lambda x : (v for v in x if v > 50)

for v in fmin(lst):
	print (v)

[75, 23, 46, 17, 53]
75
53


In [77]:
gr = lambda x,y: x if x > y else y
print (gr(10,20))


20


In [78]:
total = 0
def addValue(value):
	global total
	total += value
	return total
def subtractValue(value):
	global total
	total -= value
	return total

mylambda = lambda x, f: f(x)


print (mylambda(10, addValue))
print (mylambda(5, subtractValue))



10
5


***
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #6.7 </p>

<style>
	p {
		background-color: black;
		color: white;
		border: 1px solid grey;
		font-size: 30px;
	}
</style>
<p> Make a Lamba function that returns the average of the list element values </p>
- Example numbers = [1, 3, 2, 4, 5, 0] <br>
- listavg = getavg(numbers)  # It should be 2.5


In [79]:
# Complete your code Here. Lambda Function

getavg = lambda x: sum(x)/len(x)

numbers = [1,2,3,4,5,6,7,8,9,10]
result = getavg(numbers)
print (result)

5.5


***
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #6.8 </p>

<style>
	p {
		background-color: black;
		color: white;
		border: 1px solid grey;
		font-size: 30px;
	}
</style>
<p> Make a Lamba function that returns all the elment that are greater than the immediate right </p>
- Example numbers = [1, 3, 2, 4, 5, 0] <br>
- listavg = greaterRight(numbers)  # 3, 5


In [83]:
getGreater = lambda x: [x[i] for i in range(len(x) -1) if x[i] < x[i+1]]  # use range of 5 to prevent range error. the last variable is i+1 in the comparison.

numbers = [1,3,2,4,5,0]
result = getGreater(numbers)
print (result)

[1, 2, 4]


***
#### <p style="font-family: Arial; font-size:1.4em; background-color: yellow; color:black;"> Problem #6.9 </p>

<style>
	p {
		background-color: black;
		color: white;
		border: 1px solid grey;
		font-size: 30px;
	}
</style>
<p> Make a Lamba function that recevies two lists and determine the longer list. Return the longer list </p>
- Example n1 = [1, 3, 2, 4, 5, 0] n2 = [1,2,3]<br>
- longerlst = compareTwoList(n1, n2)  # Return n1


In [84]:
compareTwoList = lambda n1, n2: n1 if len(n1) > len(n2) else n2

n1 = [1,3,2,4,5,0]
n2 = [1,2,3]

longer = compareTwoList(n1, n2)
print (longer)

[1, 3, 2, 4, 5, 0]
